Add aten._unsafe_masked_index

isuruf opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #116491

To generate masked indexing operations that would generate
masked loads in triton code

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @rohan-varma @aakhundov

isuruf committed:
Add aten._masked_index

To generate masked indexing operations that would generate
masked loads in triton code

[ghstack-poisoned]

lezcano reviewed (changes requested):
If you are going to implement it as an ATen op, you should implement it in ATen.
You can do this two ways:
- You can get a mask per dimension and do something like do a `masked_fill(idx, 0)` for every dimension
- You can make the function less general and implement one of the following:
   - Take a a vector of pairs with length equal to the dimensions, and on each dimension take the upper and lower bounds for the indices. Then implement the function as a clamp into an `at::index`
   - Don't pass anything and just make sure that the indices are within `[-size, size-1]` for each dimension.

To choose which way to go, have a look at the uses of `ops.masked` within `lowerings.py` and see which of these (or a variation of one of those) would allow for implementing most of the pooling ops. You might need some extra args for padding and so, but yeah.

Then, we also need to implement the backward (this is the main advantage of implementing this as a ATen op vs an inductor prim. Once this PR is in a good state I can help with that part

isuruf committed:
Update on "Add aten._masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -656,6 +656,27 @@ Tensor _unsafe_index(const Tensor& self, const torch::List<c10::optional<Tensor>
+      clamped_indices.push_back(at::clamp(*index, -sizes[i], sizes[i] - 1));
+    }
+  }
+  auto result = at::index(self, clamped_indices);
```
Use `_unsafe_index`, as we know by construction that the indices are correct here.

lezcano reviewed (commented):
The backward of this function is something along the lines of (I used `std::transform` for conciseness):

```c++
Tensor _masked_index_put(const Tensor& self, const Tensor& mask, const torch::List<c10::optional<Tensor>>& indices, const Tensor& value, bool accumulate) {
  // This is the backward of _masked_index.
  // This function is not meant to be executed on eager mode.
  // We recompute the clamped indices and rely on inductor to CSE the computation
  auto clamp = [](const c10::optional<Tensor>& index, auto size) -> c10::optional<Tensor> {
      if (!index) {
          return index
      }
      // Disallow bool
      auto dtype = index->scalar_type();
      TORCH_CHECK(dtype == kLong || dtype == kInt,
                  "_masked_index found unexpected index type ", dtype);
      return at::clamp(*index, -size, size - 1);
  };

  torch::List<c10::optional<Tensor>> clamped_indices(indices.size());
  std::transform(indices.begin(), indices.end(), self.sizes.begin(), clamped_indices.begin(), clamp);
  auto masked_value = value.masked_fill(~mask, 0);
  return at::_unsafe_index_put(self, clamped_indices, masked_value, accumulate);
}
```

It should be obvious how to lower this function to masked ops.

The backward entries in `derivatives.yaml` for these two functions should be:

```c++
- name: _masked_index.Tensor(Tensor self, Tensor mask, Tensor?[] indices, Scalar Value) -> Tensor
  self: at::_masked_index_put(grad.new_zeros_symint(self.sym_sizes(), self.options()), mask, indices, grad, true)
  mask: non_differentiable
  result: auto_linear

- name: _masked_index_put(Tensor self, Tensor mask, Tensor?[] indices, Tensor values, bool accumulate=False) -> Tensor
  self: "accumulate ? grad.masked_fill(~mask, 0) : at::_masked_index_put(grad, mask, indices, zeros_like(values), false)"  
  mask: non_differentiable
  values: at::_masked_index(grad, mask, indices, 0)
  result: at::_masked_index_put(self_t, indices, values_t, accumulate)
```

I have not tested any of this, so there may be some errata really. To test that all this is correct, generate a few inputs for `_masked_index` and implement the function in the Python API and compare the values of calling `backward` for `torch._masked_index`  and your Python implementation.

peterbell10 commented on torch/_inductor/lowering.py:
```diff
@@ -3136,6 +3140,22 @@ def load_source_val():
 
+@register_lowering(aten._masked_index, type_promotion_kind=None)
+def _masked_index(self, mask, indices, fill):
+    ranges, _unsafe_index_fn = index_impl_helper(self, indices, check=False)
```
We should call the aten operator `_unsafe_masked_index` if it's not going to check the bounds.

lezcano reviewed (commented):
To abide by https://dev-discuss.pytorch.org/t/core-aten-opset-backward-forward-compatibility-policy/, we shoul dalso add decompositions for these functions.

isuruf committed:
Update on "Add aten._masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

albanD commented:
Why a
> To abide by https://dev-discuss.pytorch.org/t/core-aten-opset-backward-forward-compatibility-policy/, we shoul dalso add decompositions for these functions.

The decomposition is a nice to have but not mandatory. The wording will be make clear in the final version.

lezcano commented:
This op is just to be used by inductor and other compilers. We would make an inductor-only prim, but we want to support Autograd as well, that's why we're putting it within native_functions.yaml. This follows the pattern from `_unsafe_index`  and `_unsafe_index_put`.

isuruf committed:
Update on "Add aten._masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

albanD commented:
In that case it would be great to have detailed documentation for this function, including in what sense it is "unsafe".
Also having appropriate OpInfo based testing to ensure we maintain its behavior going forward.

isuruf committed:
Update on "Add aten._masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano commented on torch/_meta_registrations.py:
```diff
@@ -2872,6 +2872,26 @@ def nonzero_static(self, *, size: int, fill_value: int = -1):
     return self.new_empty((size, self.dim()), dtype=torch.long)
 
 
+@register_meta([aten._unsafe_masked_index])
```
If you add these checks to the decomp, you can remove the metafunction altogether.

lezcano commented on torch/_inductor/lowering.py:
```diff
@@ -3146,6 +3150,38 @@ def load_source_val():
+def _unsafe_masked_index_put(x, mask, indices, values, accumulate=False):
+    masked_value = where(mask, values, 0)
+    shape = x.get_size()
+    clamped_indices = [
+        clamp(indices[i], -shape[i], shape[i] - 1) if indices[i] else None
+        for i in range(len(indices))
+    ]
```
We can also do this via a masked operation since tl.atomic_add [accepts a mask](https://triton-lang.org/main/python-api/generated/triton.language.atomic_add.html#triton.language.atomic_add), but I don't know if inductor supports it as-is.
Can you see if it would work, and if it does, implement it with a masked index_add. Otherwise, just leave a comment and we'll look into it in a follow-up PR.

lezcano commented on tools/autograd/derivatives.yaml:
```diff
@@ -826,6 +826,17 @@
+- name: _unsafe_masked_index(Tensor self, Tensor mask, Tensor?[] indices, Scalar fill) -> Tensor
+  self: at::_unsafe_masked_index_put(grad.new_zeros_symint(self.sym_sizes(), self.options()), mask, indices, grad, true)
+  mask: non_differentiable
+  result: auto_linear
```
Almost! The correct line is:
```suggestion
  result: _unsafe_masked_index(self_t, mask, indices, 0)
```
since the derivative of the indices that we are filling in is 0 as they are constant.

lezcano reviewed (commented):
Regarding the functorch failures, you can  simply add an `OP_DECOMPOSE` to BatchRulesDecompositions.cpp and you'll get the batching rule for free. I think with these points you'll fix most of the issues

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

lezcano commented on torch/_decomp/decompositions.py:
```diff
@@ -3404,6 +3404,18 @@ def _unsafe_index(x, indices):
 
 @register_decomposition([aten._unsafe_masked_index])
 def _unsafe_masked_index(x, mask, indices, fill):
+    for index in indices:
```
You also need the checks in the backward I guess

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

zou3519 reviewed (commented):
please shout when this is ready

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

lezcano commented on torch/distributed/_tensor/dispatch.py:
```diff
@@ -274,6 +274,7 @@ def unwrap_to_op_info(
         runtime_schema_info = self.sharding_propagator.op_to_schema_info.get(
             op_call, None
         )
+        print(op_call)
```
oops

Same below.

lezcano commented on torch/_decomp/__init__.py:
```diff
@@ -440,6 +440,8 @@ def core_aten_decompositions() -> Dict[torch._ops.OperatorBase, Callable]:
             aten.unbind,
             aten.unfold_backward,
             aten.unfold_copy,
+            aten._unsafe_masked_index,
+            aten._unsafe_masked_index_put,
             aten._unsafe_index,
```
I think these should not be here. Removing these from here, should allow us to remove them from `decomps_to_exclude` as well.

lezcano replied:
If CI agrees that we can remove these, could you also remove `_unsafe_index`?

lezcano commented on test/test_fx_experimental.py:
```diff
@@ -1623,21 +1625,40 @@ def jit_infer_type(v):
+
+            idx = 0
+
+            def process_arg(arg, name):
```
why are these necessary?

isuruf replied:
The test didn't work for arguments that were lists of tensors which is needed for the indices in the `_unsafe_masked_index` operation. I fixed the test here.

lezcano commented on test/inductor/test_torchinductor.py:
```diff
@@ -1081,6 +1081,39 @@ def flip(x):
+
+        inp = [
+            torch.randn(8, device=self.device, requires_grad=True),
+            torch.tensor([True, False, True], device=self.device),
+            torch.tensor([3, 9, -2], device=self.device),
```
These two don't make sense for the backward.

isuruf replied:
I'm not sure I understand. Can you elaborate a bit?

lezcano replied:
We can't take the backward of non-floating point values, and we already test these in the forward test.

isuruf replied:
These are the indices which has to be integer values and `requires_grad` is only set for `self` which is a floating point value.

isuruf replied:
The forward test previously runs a backward test too?

lezcano reviewed (approved):
Looks very good. I just left a couple questions, but overall looks great! Thank you for the big effort, I think in this PR you have touched more than half of the subsystems in PyTorch :P

lezcano commented:
Sorry, didn't mean to approve yet.

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -656,6 +656,59 @@ Tensor _unsafe_index(const Tensor& self, const torch::List<c10::optional<Tensor>
+  std::transform(indices.begin(), indices.end(), self.sizes().begin(), clamped_indices.begin(), clamp);
+
+  auto masked_value = values.masked_fill(~mask, 0);
+  return at::_unsafe_index_put(self, clamped_indices, masked_value, accumulate);
```
This only works for `accumulate=True` otherwise you clobber with the clamped values.

lezcano replied:
The mask is, I believe, always a super set of the clamped valued. As such, when `accumulate=False`, you are just putting zeros in the clamped indices.

lezcano replied:
At any rate, if the mask is not a superset of the clamped values, this should also work really.

peterbell10 commented on tools/autograd/derivatives.yaml:
```diff
@@ -830,6 +830,17 @@
+  result: _unsafe_masked_index(self_t, mask, indices, 0)
+
+- name: _unsafe_masked_index_put(Tensor self, Tensor mask, Tensor?[] indices, Tensor values, bool accumulate=False) -> Tensor
+  self: "accumulate ? grad.masked_fill(~mask, 0) : at::_unsafe_masked_index_put(grad, mask, indices, zeros_like(values), false)"
```
When `accumulate=True` why do you zero out the self gradient? `self` should be included in the accumulated result and therefore have equal gradient.

lezcano replied:
You are right. This hasn't been caught by CI because we don't have (yet) an OpInfo for this op, and we don't really use it. But it'd be good to have for completeness.

@isuruf change `grad.masked_fill(~mask, 0)` by `grad` and we should be good here.

peterbell10 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5081,6 +5081,37 @@ def make_idx(n, m):
+        indices = [make_idx(idx_size, high, dim, d) for d in range(dim)]
+        masks = [torch.logical_and(idx >= 0, idx < self_shape[i]) for i, idx in enumerate(indices) if idx is not None]
+        mask = functools.reduce(torch.logical_and, masks)
+        args = (mask, indices, fill_value)
+        yield SampleInput(make_arg(self_shape), args=args)
```
```suggestion
        yield SampleInput(make_arg(self_shape), mask, indices, fill_value)
```

peterbell10 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5081,6 +5081,37 @@ def make_idx(n, m):
+
+        masks = [torch.logical_and(idx >= 1, idx < self_shape[i] - 1) for i, idx in enumerate(indices) if idx is not None]
+        mask = functools.reduce(torch.logical_and, masks)
+        args = (mask, indices, fill_value)
+        yield SampleInput(make_arg(self_shape), args=args)
```
```suggestion
        yield SampleInput(make_arg(self_shape), mask, indices, fill_value)
```

peterbell10 commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -16425,6 +16453,14 @@ def reference_flatten(input, start_dim=0, end_dim=-1):
            dtypes=all_types_and(torch.float16, torch.bfloat16),
            supports_out=True,
            sample_inputs_func=sample_inputs_index_reduce),
+    OpInfo('_unsafe_masked_index',
```
Missing an `OpInfo` for `_unsafe_masked_index_put`?

lezcano reviewed (commented):
So, as I was re-reading this function again, I was wondering whether we do need the mask, or we can define it to be as `and(-size <= idx <= size - 1for idx in indices)`, where the and is performed with proper broadcasting. Would this be general enough to support all the uses in pooling ops of this function, or do we ever pass in a different mask?

lezcano reviewed (commented):
So, after discussing with @peterbell10, the eager implementation of `_unsafe_masked_index_put` when `accumulate=False` is not correct as it stands, as it modifies (puts zeros) in the positions where `mask = False`, but instead, it should not touch those values. 

We could fix it, but we figured out this is not worth the effort. It'd be better to rename `_unsafe_masked_index_put` to `_unsafe_masked_index_put_accumulate`, and just implement `accumulate=True`, which is the path we care about.

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

lezcano commented on torch/_inductor/lowering.py:
```diff
@@ -3211,6 +3215,40 @@ def load_source_val():
+
+@register_lowering(aten._unsafe_masked_index_put_accumulate)
+def _unsafe_masked_index_put_accumulate(x, mask, indices, values):
+    masked_value = where(mask, values, 3)
```
Is this to catch errors? If so, can you leave a comment?

isuruf replied:
Sorry about that. I'll be more careful to not push debug stuff.

lezcano replied:
It's alright. I didn't dislike the idea of using a funny value here to catch errors. In any case, either is fine, as eventually this will be replaced by a masked atomic_add.

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

isuruf commented:
I changed to `_unsafe_masked_index_put_accumulate`. Do you think we should also remove the explicit `mask`?

lezcano commented:
You'll need to see if all the uses of the masked load in the lowerings that we want to implement with this op have a mask that's exactly equal to  what I described in https://github.com/pytorch/pytorch/pull/116491#pullrequestreview-1919522429

peterbell10 commented:
> I was wondering whether we do need the mask, or we can define it to be as `and(-size <= idx <= size - 1for idx in indices)`

This sounds a bit problematic to me. What if the decomposition uses indices in the range `[0, size - 1)` and `-1` should be masked out?

peterbell10 commented:
Test failures look like you just need to bump the tolerance for half precision in `test_torchinductor_opinfo.py`

lezcano commented:
>This sounds a bit problematic to me. What if the decomposition uses indices in the range [0, size - 1) and -1 should be masked out?

My question is, is that ever the case? We always handle negative numbers properly when doing indexing.

peterbell10 commented:
> We always handle negative numbers properly when doing indexing.

It's not a question of handling negative numbers, it's a question of loading only the data that the kernel is expected to access. Otherwise we waste precious global memory bandwidth fetching data that isn't needed.

lezcano commented:
My point is, when should `-1` be masked out when using this op in a decomp? What's an example where we want to do that.

peterbell10 commented:
Well in the top of this PR stack for example we have `torch.logical_and(idx >= -padding[i], idx < dhw[i] + padding[i])` which is an example of a more complicated mask and in the case with no padding it contains `idx >= 0`.

lezcano commented:
Then that answers my question above -> We cannot remove the mask. Thanks!

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

lezcano commented:
@wconstab there's a (simple-looking) DTensor issue with this PR, who could be the best person to help with it?

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

albanD commented:
cc @wanchaol for the DTensor issue maybe?

lezcano commented:
ping @wanchaol

wanchaol commented:
Hey @isuruf, sorry just saw this, you can simply put a `xfail` in test_dtensor_ops.py to resolve the DTensor CI failure https://github.com/pytorch/pytorch/blob/main/test/distributed/_tensor/test_dtensor_ops.py#L483

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

lezcano commented:
Once this PR passes CI, can you put one example PR on top of this where you use it in a decomposition? It'd be best to choose the simplest operation that uses this pattern.

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

isuruf committed:
Update on "Add aten._unsafe_masked_index"


To generate masked indexing operations that would generate
masked loads in triton code

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang voznesenskym EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire

[ghstack-poisoned]

peterbell10 commented on test/inductor/test_torchinductor.py:
```diff
@@ -1112,6 +1112,39 @@ def flip(x):
+            ),
+        )
+
+    def test__unsafe_masked_index_backward(self):
```
`self.common` should already be testing the backward formulas so I don't think this is necessary. You could have an explicit test for `_unsafe_masked_index_put_accumulate` though

peterbell10 commented on torch/_inductor/lowering.py:
```diff
@@ -3306,6 +3310,57 @@ def load_source_val():
+def _unsafe_masked_index(self, mask, indices, fill):
+    if torch.version.hip is not None:
+        # Avoid a triton compiler failure
+        return fallback__unsafe_masked_index(self, mask, indices, fill)
```
Do you have more logs or details about this failure?

isuruf replied:
Fails with PassManager::run failed. Log at https://ossci-raw-job-status.s3.amazonaws.com/log/22537833056

peterbell10 replied:
We can't really use this if it regresses ROCm support. From the logs you linked it looks like its the backward that's failing though, so perhaps we can still lower the forward?

BTW running CI with `torch._inductor.config.compile_threads = 1` might give a better compiler error. Some of the message gets lost when compiling in a process pool.

isuruf replied:
It's a compiler failure. https://github.com/pytorch/pytorch/pull/121801 should fix this for us.

peterbell10 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -656,6 +656,59 @@ Tensor _unsafe_index(const Tensor& self, const torch::List<c10::optional<Tensor>
+    auto dtype = index->scalar_type();
+    TORCH_CHECK(dtype == kLong || dtype == kInt,
+                "_unsafe_masked_index found unexpected index type ", dtype);
+    return at::clamp(*index, -size, size - 1);
```
Hazard here for empty inputs:
```
2024-04-11T21:17:58.2785208Z IndexError: index is out of bounds for dimension with size 0
```

It may be reasonable to just return a `full` tensor when there are no valid indices. It may be non-trivial to determine the correct output shape though.

isuruf committed:
Update

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -680,6 +680,19 @@ Tensor _unsafe_masked_index(const Tensor& self, const Tensor& mask, const torch:
+      if (!index) {
+        return size;
+      } else {
+        return index->numel();
```
I think you're assuming the index tensor here is a 1d-tensor viewed in multiple dimensions, but I don't think that's always the case.

isuruf replied:
I didn't think we would or need to support anything other than 1d-tensors as elements of the indices list. I don't know what a higher dimensional tensor here is supposed to do.

peterbell10 replied:
Here is an example of multidimensional indexing:
```python
import torch
a = torch.arange(100).reshape(10, 10)
b = torch.randn(100)
b[a] .shape # (10, 10)
```

peterbell10 replied:
A quick and dirty way to get the right output shape might be to call the meta implantation for `at::index`

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -656,6 +657,75 @@ Tensor _unsafe_index(const Tensor& self, const torch::List<c10::optional<Tensor>
+    };
+    std::transform(indices.begin(), indices.end(), self.sizes().begin(), new_sizes.begin(), compute_new_size);
+
+    auto result = at::meta::index(self.new_empty(new_sizes), clamped_indices);
+    return self.new_full(result.sizes(), fill);
```
Since we don't actually read the indices, you can move this early return before we clamp the indices.

peterbell10 commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -656,6 +657,75 @@ Tensor _unsafe_index(const Tensor& self, const torch::List<c10::optional<Tensor>
+    auto dtype = index->scalar_type();
+    TORCH_CHECK(dtype == kLong || dtype == kInt,
+                "_unsafe_masked_index found unexpected index type ", dtype);
+    return at::clamp(*index, -size, size - 1);
```
The same hazard exists for index put. You can just return `self.clone()` if self is an empty tensor.

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf commented:
Any idea how to fix the compilation failures?

peterbell10 commented:
Hrm looks like meta dispatch isn't available on mobile:
https://github.com/pytorch/pytorch/blob/70ad64e8a644d478228aa740e03f65d6153c4074/cmake/Codegen.cmake#L87

I guess you can just wrap references to it in `#ifndef C10_MOBILE` and otherwise raise a runtime error if it's called on mobile.

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf commented:
This is ready for another review.

lezcano commented on test/inductor/test_torchinductor_opinfo.py:
```diff
@@ -369,6 +369,8 @@ def wrapper_noop_set_seed(op, *args, **kwargs):
     },
     ("std_mean.unbiased", "cuda", f16): {"reference_in_float": True},
     ("uniform", "cuda"): {"reference_in_float": True},
+    ("_unsafe_masked_index_put_accumulate", "cuda", f16): {"atol": 1e-4, "rtol": 0.01},
+    ("_unsafe_masked_index_put_accumulate", "cpu", f16): {"atol": 1e-4, "rtol": 0.01},
```
why don't we have hte same in `index_put`?

isuruf replied:
I think it's because the test input samples are slightly different.

lezcano reviewed (approved):
Massive effort @isuruf! This LGTM, but let's also wait for @peterbell10 to have a look

peterbell10 commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1974,7 +1974,10 @@ def load(self, name: str, index: sympy.Expr):
             else:
-                line = f"tl.load({var} + ({indexing.index_str}), {indexing.mask_str}{ep}{other})"
+                if torch.version.hip is not None and indexing.mask_str != "None":
+                    line = f"tl.load({var} + ({indexing.index_str}), ({indexing.mask_str}).to(tl.int1){ep}{other})"
```
NIT: It would be better to convert to boolean in `ops.masked` rather than at the point where the mask is used.

isuruf committed:
Update

[ghstack-poisoned]

isuruf commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -656,6 +656,82 @@ Tensor _unsafe_index(const Tensor& self, const torch::List<c10::optional<Tensor>
+      // We use a hack here since we do not have a method to get the
+      // correct size of the tensor. (except with meta impl which is
+      // not available on mobile builds)
+      std::vector<int64_t> new_sizes(self.dim());
+      auto compute_new_size = [](const c10::optional<Tensor>& index, auto size) -> int64_t {
+          if (index && size == 0) {
+              return 1;
+          } else {
+              return size;
+          }
+      };
+      std::transform(indices.begin(), indices.end(), self.sizes().begin(), new_sizes.begin(), compute_new_size);
+      auto result = self.new_full(new_sizes, fill);
+      return at::_unsafe_index(result, clamped_indices);
```
>  To be passed to _unsafe_index It should be like the size of self, but with zero-dimensions expanded to 1.

I'm not sure I understand. This is what I do.

peterbell10 replied:
Ah this was a pending comment and I guess it's out of date. Indeed this looks fine.

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

lezcano commented on tools/autograd/derivatives.yaml:
```diff
@@ -830,6 +830,17 @@
+  mask: non_differentiable
+  result: _unsafe_masked_index(self_t, mask, indices, 0)
+
+- name: _unsafe_masked_index_put_accumulate(Tensor self, Tensor mask, Tensor?[] indices, Tensor values) -> Tensor
```
not quite. `index_add` does that but along just one dimension

peterbell10 replied:
Ah true

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf commented:
@pytorchbot merge

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf commented:
>  inductor/test_torchinductor_opinfo.py::TestInductorOpInfoCPU::test_comprehensive__unsafe_masked_index_cpu_bool

This test failure is real. I can reproduce locally on linux (It's strange that this error came up only on macos run)

The failure seems to come from special casing of bool at https://github.com/pytorch/pytorch/blob/91bf952d10e9524a9b078900d9807efa5d252f5c/torch/_inductor/codegen/cpp.py#L2077-L2085

isuruf committed:
Update

[ghstack-poisoned]

lezcano commented:
>The failure seems to come from special casing of bool

Is it an easy fix? If so, can you prepend a PR with the fix?

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf commented:
> Is it an easy fix? If so, can you prepend a PR with the fix?

Not sure. I opened https://github.com/pytorch/pytorch/pull/126406, but it's most probably not the correct solution.

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

peterbell10 commented:
Given this is a pre-existing issue in the C++ backend I think it's fine to open an issue for the Intel folks to look at and just skip the test for this PR.

isuruf committed:
Update

[ghstack-poisoned]

peterbell10 commented on test/test_mps.py:
```diff
@@ -139,8 +139,8 @@ def mps_ops_grad_modifier(ops):
-        '_unsafe_masked_index': [torch.int8, torch.uint8, torch.float16, torch.int16, torch.int64],
-        '_unsafe_masked_index_put_accumulate': [torch.int8, torch.uint8, torch.float16, torch.int16, torch.int64],
+        '_unsafe_masked_index': [torch.bool, torch.int8, torch.uint8, torch.float16, torch.int16, torch.int64],
+        '_unsafe_masked_index_put_accumulate': [torch.bool, torch.int8, torch.uint8, torch.float16, torch.int16, torch.int64],
```
How do integer types end up in the backward pass?

isuruf replied:
I'm not sure I understand the question. MPS accumulate supports only `float32` and `int32`

peterbell10 replied:
The comment says "atomic operations in the backward pass" but you can't get the gradient of an integer tensor. So something isn't right here.

isuruf replied:
Thanks. I misread the test failures. Ready for a review now.

isuruf committed:
Update

[ghstack-poisoned]

isuruf commented:
>  I can reproduce locally on linux (It's strange that this error came up only on macos run)

The only CI with non-AVX512 hardware that this test runs on are the macos runners.

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

lezcano commented:
Deleted the merge just in case Peter has more comments, but this LGTM. Feel free to merge if you agree, Peter.

peterbell10 commented on test/inductor/test_torchinductor.py:
```diff
@@ -10845,7 +10872,10 @@ def fn(x, n):
             # load should be masked
-            self.assertTrue("tl.load(in_ptr0 + (tmp0), xmask" in code)
+            self.assertTrue(
+                "tl.load(in_ptr0 + (tmp0), xmask" in code
+                or "tl.load(in_ptr0 + (tmp0), (xmask).to(tl.int1)" in code
```
NIT: I think this change is stale, since the `.to` was moved from the load itself to `ops.masked`.

isuruf committed:
Update

[ghstack-poisoned]

isuruf commented:
@pytorchbot, merge

isuruf commented:
@pytorchbot merge

PaliC commented:
@pytorchbot help

PaliC commented:
@pytorchbot revert -m "breaking onnx tests" -c "nosignal"

isuruf commented:
@PaliC, any idea on how to reproduce the failure?

peterbell10 commented:
@isuruf here is the hud entry
https://hud.pytorch.org/pytorch/pytorch/commit/f03f8bc901a6c9038308a6353e8d280f4b5628f5

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf commented:
Is it possible to run the checks that ran on main branch here?

isuruf committed:
Update

[ghstack-poisoned]

isuruf commented:
@bowenbao @thiagocrepaldi @wschin I can reproduce the failure locally, but I don't know what to do here. Any ideas?

lezcano commented:
Is it possible to skip the test?

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

titaiwangms commented:
> @BowenBao @thiagocrepaldi @wschin I can reproduce the failure locally, but I don't know what to do here. Any ideas?

Hi @isuruf LMK if you still have issue xfail/skip the onnx tests. It looks good to me.

isuruf commented:
@titaiwangms, yes, I'm xfailing the test cases here for now.

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

isuruf committed:
Update

[ghstack-poisoned]

isuruf committed:
Update

[ghstack-poisoned]

isuruf commented:
@pytorchbot merge -i

isuruf committed:
Update

[ghstack-poisoned]
